src/main/python/deep_qa/layers/noisy_or.py

```diff
@@ -0,0 +1,89 @@
+        p *= K.cast(p >= 0., K.floatx())
+        leaky_zeros_mask = K.epsilon() * K.cast(K.equal(p, K.zeros_like(p)), 'float32')
+        p = p + leaky_zeros_mask
+        # Clip values greater to 1 to 1
+        p *= K.cast(p <= 1., K.floatx())
+        leaky_ones_mask = K.cast(K.equal(p, K.zeros_like(p)), K.floatx())
+        p += leaky_ones_mask
+        return p
+
+
+class NoisyOr(Layer):
+    """
+    Calculate the noisy-or probability across a given tensor axis based on
+    noisy-or equation:
+        p(x) = 1 - Prod_(i=1:N)[1 - q * p(x|y_n)]
+    where q is the noise parameter
+
+    Parameters
+    ----------
+    inputs: Tensor, dim=(b, ..., N, ...), where N is the number of y's in the above equation (i.e. the number of
+            probabilities being combined in the product), in the dimension
+            corresponding to the specified axis
+
+    mask: Tensor or None, optional (default=None)
+        Tensor dimensions should match the dimensions of the input tensor.
+
+    Returns
+    -------
+    output: Tensor
+        The combined probability of the even happening, given the parent events
+        Output dims are one less than input dims, with the lost dimension corresponding to the specified axis.
+    """
+    def __init__(self, axis=-1, param_init='uniform', name="noisy_or", noise_param_constraint=None, **kwargs):
+        self.axis = axis
+        self.param_init = param_init
+        self.supports_masking = True
+        self.name = name
+        # The noisy-or equation includes a noise parameter (q) which is learned during training
+        self.noise_parameter = None
+        self.noise_param_constraint = noise_param_constraint
+        super(NoisyOr, self).__init__(**kwargs)
+
+    def build(self, input_shape):
+        # Add the trainable weight variable for the noise parameter
+        self.noise_parameter = self.add_weight(shape=(1,),
+                                               name=self.name + '_noise_param',
+                                               initializer=self.param_init,
+                                               regularizer=L1L2Regularizer(l2=0.001),
+                                               constraint=self.noise_param_constraint,
+                                               trainable=True)
+        super(NoisyOr, self).build(input_shape)
+
+    def get_output_shape_for(self, input_shape):
+        if self.axis == -1:
+            return input_shape[0:-1]
+        return input_shape[:self.axis - 1] + input_shape[self.axis:]
+
+    @overrides
+    def compute_mask(self, inputs, mask=None):
+        # pylint: disable=unused-argument
+        # We do not need a mask beyond this layer.
+        # (probably right)
+        if mask is not None:
+            return K.any(mask, axis=-1)
+        return None
+
+    def call(self, x, mask=None):
+        # dim: (b, ..., num_probs, ...)
+        noisy_probs = K.eval(self.noise_parameter) * x
```

Comment:
I was just worried about `K.eval()` breaking the computation graph, so the weight doesn't get updated correctly.  Can you verify that the weight is the same after training with and without this in tensorflow?